File: Data Structures & Algorithms/jump-game/submission-0.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int maxReach = 0;   // farthest index we can reach so far

        for (int i = 0; i < nums.size(); i++) {
            // If current index is beyond what we can reach
            if (i > maxReach) {
                return false;
            }

            // Update farthest reachable index
            maxReach = max(maxReach, i + nums[i]);
        }

        return true;
    }
};

File: Data Structures & Algorithms/house-robber-ii/submission-0.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        def helper(house):
            i, j = 0, 0
            for n in house:
                new_rob = max(j, i +n)
                i = j
                j = new_rob
            return j

        return max(helper(nums[:-1]), helper(nums[1:]))

        